### PR TITLE
Be able to define global primary key type

### PR DESCRIPTION
**Pull request by girardinsamuel**
Quick try to implements #122 

@josephmancuso @hamzzy
Is something like this making sense ? doable ?

### CODE REVIEW

**Comment by josephmancuso** [comment]
I'm not entirely sure what that ticket is asking for to be honest. Either way I don't think this would be the correct approach here. I think the way to go would be do make a global scope and inherit it on the model.

**Comment by josephmancuso** [comment]
I think the scope would look something like this:

```python
class UUIDPrimary(BaseScope):
    """Global scope class to add uuid primary key"""

    def on_boot(self, builder):
        builder.set_global_scope(
            "_uuid_primary", self.set_uuid_create, action="insert"
        )

    def on_remove(self, builder):
        pass

    def set_uuid_create(self, builder):
        builder._creates.update(
            {
                builder._model.__primary_key__: uuid4(), # On insert, generate a uuid4 key right before it will try to create this record
            }
        )
```

then make a mixin to register it to the model and then you can have something like this:

@rfschubert would a syntax like this make this easier for you?

```python
class Client(Model, SoftDeletesMixin, UUIDPrimaryMixin, ...):
    pass
```

**Comment by girardinsamuel** [comment]
Yeah it looks cleaner with a mixin ! But we have to add the mixin to every model though. Definitely doable because it is not very complex !

**Comment by josephmancuso** [comment]
@girardinsamuel yes or have a base model and just have all your models inherit that model. Maybe that's not such a terrible idea to have that right out of the box with Masonite.

I'm just not sure if this achieves @rfschubert goal.

**Comment by girardinsamuel** [comment]
We will wait for his answer but as a Django user I've come across the same problem of rewriting this uuid line and for me this solves the problem

**Comment by josephmancuso** [comment]
@girardinsamuel oh perfect. Yeah I think it would work too but I thought maybe I didn't understand the original feature request. Ok if you think the global scope approach is good I don't see why we shouldn't make it :) I think it would be a cool scope even if it doesn't solve the problem lol. we might need to support more than just the primary key as guid though. Maybe we do something like `__uuids__ = ['object_id', 'transaction_id']` and we just loop through these and create uuids for each. Default to uuid4 but can also set an option in `__uuid_version__ = '4'`. Once everything is set we can put all the if checks in the `set_uuid_create` method of the scope

**Comment by girardinsamuel** [comment]
@josephmancuso Like this idea, would be more flexible 👍 !
But from a Django perspective, it's a bit strange for me not be able to do this in the migrations :

```python
with self.schema.create('users') as table:
            table.uuid('id').primary()  # uuid as PK
            table.string('name')
            table.uuid('transaction_id') # other uuid field
            # or table.uuid3('transaction_id') ...
```

Instead we need to do if I got it correctly:

```python
# User.py
class User(Model, UUIDFieldsMixin):
    __uuids__ = ['id', 'transaction_id']
    __uuid_version__ = 4

# create_user_table_migration.py
with self.schema.create('users') as table:
            table.char('id')  # ?
            table.string('name')
            table.char('transaction_id') # ?
```
Okay for me too but as I am used to "the first one in Django", it's not intuitive.

**Comment by josephmancuso** [comment]
@girardinsamuel yeah I agree. We can add support for a uuid in the schema like you have in this PR. That's fine for me. We would have to map uuid type to CHAR and then set the length default to 36 in the uuid schema method so that something like CHAR(36) will be used when you do table.uuid(..)

**Comment by girardinsamuel** [comment]
Okay cool, I will try to complete and clean up this PR (and add some tests !) Time to put hands on the ORM code 😄

File: src/masoniteorm/schema/platforms/MySQLPlatform.py [hunk 1]
```diff
@@ -15,7 +15,7 @@ class MySQLPlatform(Platform):
         "small_integer": "SMALLINT",
         "medium_integer": "MEDIUMINT",
         "increments": "INT UNSIGNED AUTO_INCREMENT PRIMARY KEY",
-        "uuid_increments": "CHAR PRIMARY KEY",
+        "uuid": "CHAR(36)",
```
Comment (josephmancuso) [inline src/masoniteorm/schema/platforms/MySQLPlatform.py]: This should be CHAR. I would not hard code a length here

File: src/masoniteorm/schema/Blueprint.py [hunk 2]
```diff
@@ -428,7 +443,7 @@ def primary(self, column):
         Returns:
             self
         """
-        self.table.set_primary_key([column])
+        self.table.set_primary_key(column)
```
Comment (girardinsamuel) [inline src/masoniteorm/schema/Blueprint.py:446]: @josephmancuso This was causing a bug so I fixed it, is it correct ?

File: src/masoniteorm/schema/platforms/PostgresPlatform.py [hunk 3]
```diff
@@ -15,6 +15,7 @@ class PostgresPlatform(Platform):
         "small_integer": "SMALLINT",
         "medium_integer": "MEDIUMINT",
         "increments": "SERIAL UNIQUE",
+        "uuid": "CHAR(36)",
```
Comment (josephmancuso) [inline src/masoniteorm/schema/platforms/PostgresPlatform.py]: same thing here

File: src/masoniteorm/schema/platforms/SQLitePlatform.py [hunk 4]
```diff
@@ -16,6 +16,7 @@ class SQLitePlatform(Platform):
         "small_integer": "SMALLINT",
         "medium_integer": "MEDIUMINT",
         "increments": "INTEGER PRIMARY KEY",
+        "uuid": "CHAR(36)",
```
Comment (josephmancuso) [inline src/masoniteorm/schema/platforms/SQLitePlatform.py]: same thing here

File: src/masoniteorm/schema/Blueprint.py [hunk 5]
```diff
@@ -85,8 +85,8 @@ def increments(self, column, nullable=False):
         )
         return self
 
-    def uuid_increments(self, column, nullable=False):
-        """Sets a column to be the UUID primary key representation for the table.
+    def uuid(self, column, nullable=False):
```
Comment (josephmancuso) [inline src/masoniteorm/schema/Blueprint.py]: You should accept a `length=36` here.

File: src/masoniteorm/schema/Blueprint.py [hunk 6]
```diff
@@ -97,9 +97,7 @@ def uuid_increments(self, column, nullable=False):
         Returns:
             self
         """
-        self._last_column = self.table.add_column(
-            column, "uuid_increments", nullable=nullable
-        )
+        self._last_column = self.table.add_column(column, "uuid", nullable=nullable)
```
Comment (josephmancuso) [inline src/masoniteorm/schema/Blueprint.py]: add pass the `length=length` argument here

**Comment by girardinsamuel** [comment]
Progress so far
- [x]  add "uuid" column in Blueprint, it's working (test_can_add_uuid_column)
- [ ]  handle auto generation of uuid when column is PK
- [ ]  add uuid version configuration

**Review by josephmancuso — CHANGES_REQUESTED** [review]
This is perfect except you shouldn't hardcode the length in the platform. If you pass the length through the schema builder via `uuid('col', length=36)` it will compile down to `CHAR(36)`.

**Comment by girardinsamuel** [comment]
Now the ideal would be to add `UUIDPrimaryKeyScope` to `_global_scopes` when a primary key constrained column is using uuid. Then the for loop on scopes already present on builder would be able to generate the uuid when inserting a new record.

But I don't know how/when to "link" `table.uuid().primary()` to `builder.set_global_scope(UUIDPrimaryKeyScope())` : @josephmancuso do you have an idea ? Am I on the right track 🤔 ?

File: src/masoniteorm/schema/Blueprint.py (hunk 2 above)
Comment (josephmancuso) [reply src/masoniteorm/schema/Blueprint.py:446]: Uhh yeah looks right. I have to check the table tests

File: src/masoniteorm/query/QueryBuilder.py [hunk 7]
```diff
@@ -349,6 +348,12 @@ def create(self, creates={}, query=False, id_key="id", **kwargs):
             creates = kwargs
 
         self.set_action("insert")
+
+        # here is an not a right place to do that but somewhere we should do this:
+        # TODO: if pk column is of type uuid, (how?)
+        # then add scope
+        self.set_global_scope(UUIDPrimaryKeyScope())
```
Comment (josephmancuso) [inline src/masoniteorm/query/QueryBuilder.py]: This is done inside the mixin. The mixin has a boot where you call this method

File: src/masoniteorm/scopes/UUIDPrimaryKeyScope.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import uuid
+from .BaseScope import BaseScope
+
+
+class UUIDPrimaryKeyScope(BaseScope):
```
Comment (josephmancuso) [inline src/masoniteorm/scopes/UUIDPrimaryKeyScope.py:5]: It's a little confusing but now that we have a global scope we register it to a mixin and then the mixin gets inherited by the model

File: src/masoniteorm/scopes/UUIDPrimaryKeyScope.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+            "_uuid", self.set_uuid_create, action="insert"
+        )
+
+    def on_remove(self, builder):
+        pass
+
+    def set_uuid_create(self, builder):
+        builder._creates.update(
+            {
+                # TODO: later how to fetch uuid version here ?
```
Comment (josephmancuso) [inline src/masoniteorm/scopes/UUIDPrimaryKeyScope.py]: Builder._model will give you the model. This is where you can check if it has any version attributes. You can set attributes on the mixin class which will get inherited down to the model when you inherit. You can default to 4

**Review by josephmancuso — CHANGES_REQUESTED** [review]
This is perfect. The last missing piece of the puzzle is the Mixin class. You have your global scope created here. Now we need a mixin file like UUIDPrimaryKeyMixin or something. The mixin file will import this global scope you created here. 

When the model boots it will grab all parent classes that end with "Mixin" and call the boot method on it. In that boot method we call that builder.set_global_scope() method. 

I'll link an example. We do this same exact thing with time stamps.

**Comment by josephmancuso** [comment]
@girardinsamuel check the TimeStampMixin class as an example

**Comment by josephmancuso** [comment]
This mixin class is what the user will actually import and inherit to their models.

**Comment by girardinsamuel** [comment]
> This is perfect. The last missing piece of the puzzle is the Mixin class. You have your global scope created here. Now we need a mixin file like UUIDPrimaryKeyMixin or something. The mixin file will import this global scope you created here.
> 
> When the model boots it will grab all parent classes that end with "Mixin" and call the boot method on it. In that boot method we call that builder.set_global_scope() method.
> 
> I'll link an example. We do this same exact thing with time stamps.

Thanks for the tips ! I was aware of the Mixin but was still looking forward to do it without it but I guess that's better with a Mixin rather than adding code where it should not belong. So we have a tradeoff between the two solutions 👍 
Now we can do:
```python
# migration
with self.schema.create('users') as table:
            table.uuid('id')
            table.primary('id')
            table.string('name')
            table.uuid('transaction_id')

# User.py
class User(Model, UUIDPrimaryKeyMixin):
    __uuid_version__ = 3
```

I am currently adding more tests on default global scopes, in this PR

**Review by josephmancuso — APPROVED** [review]
perfect. Solves the problem quite nicely.

**Pull request merged**